plutus-contract/src/Plutus/Contract/StateMachine/ThreadToken.hs

```diff
@@ -0,0 +1,81 @@
+{-# LANGUAGE DataKinds             #-}
+{-# LANGUAGE DeriveAnyClass        #-}
+{-# LANGUAGE DerivingStrategies    #-}
+{-# LANGUAGE FlexibleContexts      #-}
+{-# LANGUAGE MonoLocalBinds        #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE NamedFieldPuns        #-}
+{-# LANGUAGE NoImplicitPrelude     #-}
+{-# LANGUAGE OverloadedStrings     #-}
+{-# LANGUAGE TemplateHaskell       #-}
+{-# LANGUAGE TypeApplications      #-}
+{-# LANGUAGE TypeOperators         #-}
+{-# LANGUAGE ViewPatterns          #-}
+{-# OPTIONS_GHC -fno-ignore-interface-pragmas #-}
+module Plutus.Contract.StateMachine.ThreadToken where
+
+import           PlutusTx.Prelude     hiding (Monoid (..), Semigroup (..))
+
+import           Data.Aeson           (FromJSON, ToJSON)
+import           GHC.Generics         (Generic)
+import           Ledger               (CurrencySymbol, TxOutRef (..))
+import qualified Ledger.Contexts      as V
+import           Ledger.Scripts
+import qualified Ledger.Typed.Scripts as Scripts
+import           Ledger.Value         (TokenName (..), Value (..))
+import qualified Ledger.Value         as Value
+import qualified PlutusTx
+import qualified PlutusTx.AssocMap    as Map
+import qualified Prelude              as Haskell
+
+data ThreadToken = ThreadToken
+    { ttOutRef         :: TxOutRef
+    , ttCurrencySymbol :: CurrencySymbol
+    }
+    deriving stock (Haskell.Eq, Haskell.Show, Generic)
+    deriving anyclass (ToJSON, FromJSON)
+
+PlutusTx.makeLift ''ThreadToken
+
+validate :: TxOutRef -> (ValidatorHash, Bool) -> V.ScriptContext -> Bool
+validate (TxOutRef refHash refIdx) (vHash, burn) ctx@V.ScriptContext{V.scriptContextTxInfo=txinfo} =
```

Comment:
Doesn't this let you burn the token at any time? That seems too strong. Someone could burn the token in the middle of the machine operation. Something needs to keep it from being burned until the end. Although maybe that's handled by the state machine's balancing operations? Seems subtle, and worth a note. TBH, the whole setup is worth a note explaining how it all fits together!

Reply:
This one I'm really not sure about. I don't feel at home enough with all of this yet to be able to think it all through. @j-mueller what do you think?

Reply:
Could you write down what you *think* happens? Then we can see if we all agree.

Reply:
The state machine validator prevents burning the token in the middle of the machine operation, even if the MPS would allow it (so you can't spend the state machine output to get the token out)